beacon-chain/sync/initial-sync/fsm.go

```diff
@@ -0,0 +1,232 @@
+	e, ok := sm.events[event]
+	if !ok {
+		e = &stateMachineEvent{
+			name: event,
+		}
+		sm.events[event] = e
+	}
+	action := &stateMachineAction{
+		state:     state,
+		handlerFn: fn,
+	}
+	e.actions = append(e.actions, action)
+	return e
+}
+
+// trigger invokes the event on a given epoch's state machine.
+func (sm *stateMachine) trigger(name eventID, epoch uint64, data interface{}) error {
+	event, ok := sm.events[name]
+	if !ok {
+		return fmt.Errorf("event not found: %v", name)
+	}
+
+	ind := sm.findEpochState(epoch)
+	if ind >= len(sm.epochs) {
+		return fmt.Errorf("state for %v epoch not found", epoch)
+	}
+
+	for _, action := range event.actions {
+		if action.state != sm.epochs[ind].state {
+			continue
+		}
+		state, err := action.handlerFn(sm.epochs[ind], data)
+		if err != nil {
+			return err
+		}
+		sm.epochs[ind].setState(state)
+	}
+
+	return nil
+}
+
+// addEpochState allocates memory for tracking epoch state.
+func (sm *stateMachine) addEpochState(epoch uint64) {
+	state := &epochState{
+		epoch:   epoch,
+		state:   stateNew,
+		blocks:  make([]*eth.SignedBeaconBlock, 0, allowedBlocksPerSecond),
+		updated: time.Now(),
+	}
+	sm.epochs = append(sm.epochs, state)
+}
+
+// removeEpochState frees memory of processed epoch.
+func (sm *stateMachine) removeEpochState(epoch uint64) error {
+	ind := sm.findEpochState(epoch)
+	if ind >= len(sm.epochs) {
+		return fmt.Errorf("state for %v epoch not found", epoch)
+	}
+	sm.epochs[ind].blocks = nil
+	sm.epochs[ind] = sm.epochs[len(sm.epochs)-1]
+	sm.epochs = sm.epochs[:len(sm.epochs)-1]
+	return nil
+}
+
+// findEpochState returns index at which state.epoch = epoch, or len(epochs) if not found.
+func (sm *stateMachine) findEpochState(epoch uint64) int {
+	for i, state := range sm.epochs {
+		if epoch == state.epoch {
+			return i
+		}
+	}
+	return len(sm.epochs)
+}
+
+// isLowestEpochState checks whether a given epoch is the lowest for which we know epoch state.
+func (sm *stateMachine) isLowestEpochState(epoch uint64) bool {
+	if ind := sm.findEpochState(epoch); ind == len(sm.epochs) {
+		return false
+	}
+	for _, state := range sm.epochs {
+		if epoch > state.epoch {
+			return false
+		}
+	}
+	return true
+}
+
+// highestEpochSlot returns slot for the latest known epoch.
+func (sm *stateMachine) highestEpochSlot() (uint64, error) {
+	if len(sm.epochs) == 0 {
+		return 0, errors.New("no epoch states exist")
+	}
+	highestEpochSlot := sm.epochs[0].epoch
+	for _, state := range sm.epochs {
+		if state.epoch > highestEpochSlot {
+			highestEpochSlot = state.epoch
+		}
+	}
+	return highestEpochSlot, nil
+}
+
+// String returns human readable representation of a state.
+func (sm *stateMachine) String() string {
+
+	return fmt.Sprintf("%v", sm.epochs)
+}
+
+// String returns human-readable representation of an epoch state.
+func (es *epochState) String() string {
+	return fmt.Sprintf("%v:%v", es.epoch, es.state)
```

Comment:
```suggestion
	return fmt.Sprintf("%d:%d", es.epoch, es.state)
```

Reply:
While epoch is a number indeed, `es.state` (of type `stateID`) has `String()` method itself. Will change the first one to `%d`, thank you.